### ISSUE TITLE
Use `tool.poetry.dependencies.python` from `pyproject.toml` to constrain the Python version

### ISSUE THREAD

**Issue by joshuatcasey** [paketo-buildpacks/poetry#17]
The `poetry` buildpacks do not seem to take the user-requested python version into account.

If you modify this version range, the integration tests fail: https://github.com/paketo-buildpacks/python-start/blob/faec2a13a9a05dc056701fc8f0238abde08bec69/integration/testdata/poetry-dep-only/pyproject.toml#L8. 

For example, I might expect `python = "3.6" to select an appropriate `3.6.x` version from the `cpython` buildpack, but instead it fails.

Note that `tool.poetry.dependencies.python` is required: https://python-poetry.org/docs/pyproject/#dependencies-and-dev-dependencies

**Comment by robdimsdale** [comment]
Agreed - this seems like an oversight. The poetry buildpack (i.e. this buildpack) should parse the `pyproject.toml` file and return the python version declared inside it as part of its `detect` phase.

I'd expect this to look similar to how the `go-mod-vendor` buildpack parses `go.mod` for the golang version: https://github.com/paketo-buildpacks/go-mod-vendor/blob/f07f3262b8a6fa8a565ba59bb8edf5f68de353c6/detect.go#L24

I'm currently validating the integration test failure you describe, and, assuming I see the same failure, I think this is high priority for us to fix.

**Comment by robdimsdale** [comment]
So, I think that the integration tests fail because the `pyproject.toml` is out of sync with the `poetry.lock` file. That's an issue, but it's a separate issue to explore. I don't yet know if the buildpack should recreate the lock file or if that is a responsibility of the application developer.

Once I remove the `poetry.lock` file, and change the version of python in the `pyproject.toml` to `3.6` (from `^3.6`) I see that `pack build` (and the integration tests which run `pack build`) succeeds in building, but does not use the version of python specified in the `pyproject.toml`:

```
[builder] Paketo CPython Buildpack 0.7.5
[builder]   Resolving CPython version
[builder]     Candidate version sources (in priority order):
[builder]                 -> ""
[builder]       <unknown> -> ""
[builder]
[builder]     Selected CPython version (using ): 3.8.12
```

So, this is still a bug, but it's less catastrophic that we first thought. The application builds with the wrong version of python. I do think it's a priority to fix as I could imagine situations where some applications cannot build with the most recent version of python, or the application developers actually want to respect the version of python in `pyproject.toml`.

**Comment by robdimsdale** [comment]
And now I no longer see failures in `pack build` / integration tests when I restore the `poetry.lock` but change the version of `python` in `pyproject.toml`. 😕

**Comment by joshuatcasey** [comment]
Sounds like we could explore the various methods of setting the `python` and/or `poetry` versions and make sure of the right priority. 

* Can we assume that `pyproject.toml` and `poetry.lock` are always in sync?
* Since `tool.poetry.dependencies.python` is required, should we raise an error when `BP_CPYTHON_VERSION` is also set?
* etc

**Comment by robdimsdale** [comment]
> Can we assume that `pyproject.toml` and `poetry.lock` are always in sync?

Sure - at least for the scope of this issue.

> Since `tool.poetry.dependencies.python` is required, should we raise an error when `BP_CPYTHON_VERSION` is also set?

I don't think we need to do this. I think this will naturally fail because the lifecycle will fail to find a build order because the version of `python` that is required by `poetry` will be inconsistent with the version provided by `python` via the environment variable. Again, taking our cues from the `go` buildpack - this seems to be how it is handled there.

**Comment by robdimsdale** [comment]
So I think the steps are:

* Add an integration test with a fixture that has a `pyproject.toml` with a version of python that is not the default (i.e. anything before `3.8`). See that the test fails because the version of `python` provided is the default from the `python` buildpack (currently `3.8`)
* Introduce a `pyproject.toml` parser that pulls the python version from `tool.poetry.dependencies.python`. Invoke this inside `detect.go`.
  * We know this field is required by `poetry` so it's ok to return the underlying error if the field is missing or cannot be parsed.
  * If the file is not present at all, return the `packit.Fail` sentinal error (this is the current behavior).
  * Use [go-mod-vendor](https://github.com/paketo-buildpacks/go-mod-vendor/blob/main/go_mod_parser.go) and [poetry-run](https://github.com/paketo-buildpacks/poetry-run/blob/main/detect.go#L32) for inspiration on parsing files and getting semantic versions.
* I don't think that there is any behavior change for `build.go` - regardless of the specified `python` version, the `build` phase should always download and install `poetry`.
  * Return any underlying failures (e.g. due to incompatible versions of `poetry` and `python`) that arise during the `build` phase.

**Comment by robdimsdale** [comment]
Also https://github.com/paketo-buildpacks/pipenv-install/pull/11 is another example of other buildpacks with this behavior.

**Comment by joshuatcasey** [comment]
This might be more tricky than anticipated.

`pyproject.toml` can contain a version range, such as `^3.6`, and `poetry.lock` does not specify an exact python version.

**Can the `poetry` buildpack be aware of which python versions are available and make the appropriate selection?**

Side note: I haven't really checked that the version ranges [supported by Poetry](https://python-poetry.org/docs/dependency-specification/) align with the version ranges [supported by masterminds/semver](https://github.com/Masterminds/semver).

**Comment by joshuatcasey** [comment]
Probably this functionality should be moved upstream to the `cpython` buildpack, so that it becomes aware of poetry so that it can make an appropriate version selection.

https://github.com/paketo-buildpacks/cpython/blob/b1e9ef1cf7d4107a61a3a89793c3e34129b57a43/constants.go#L14-L19

**Comment by joshuatcasey** [comment]
Maybe things are good. Looks like `cpython` buildpack (and others) can support a semver range in the build plan. 

https://github.com/paketo-buildpacks/packit/blob/01b7e947a72b1fdbeb5cdd5dfcbeeae25bef884b/postal/service.go#L61-L67
